tests/uefistored/test_secure_boot.py

```diff
@@ -0,0 +1,279 @@
+import os
+import pytest
+import subprocess
+from lib.common import wait_for
+from lib.efi import EFIAuth, EFI_AT_ATTRS, EFI_AT_ATTRS_BYTES, EFI_GUID_STRS
+
+VM_SECURE_BOOT_FAILED = 'VM_SECURE_BOOT_FAILED'
+GUEST_CERT_DIR = '/security/certs'
+
+
+def check_sb_failed(vm):
+    vm.start()
+    wait_for(
+        lambda: vm.get_messages(VM_SECURE_BOOT_FAILED),
+        'Wait for message %s' % VM_SECURE_BOOT_FAILED,
+    )
+
+    # If there is a VM_SECURE_BOOT_FAILED message and yet the OS still
+    # successfully booted, this is a uefistored bug
+    assert not vm.is_ssh_up(), 'The OS booted when it should have failed'
+
+
+def check_sb_succeeded(vm):
+    vm.start()
+    vm.wait_for_vm_running_and_ssh_up()
+    assert not vm.get_messages(VM_SECURE_BOOT_FAILED)
+
+
+def sign_efi_bins(vm, db):
+    '''Boots the VM if it is halted, signs the bootloader, and halts the
+    VM again (if halted was its original state).
+    '''
+    shutdown = not vm.is_running()
+    if shutdown:
+        vm.start()
+        vm.wait_for_vm_running_and_ssh_up()
+
+    print('> Sign bootloader')
+    vm.sign_efi_bins(db)
+
+    if shutdown:
+        vm.shutdown(verify=True)
+
+
+def install_auths(vm, auths, use_xapi=False):
+    for auth in auths:
+        print('> Setting {}'.format(auth.name))
+        dest = '/usr/share/uefistored/%s.auth' % auth.name
+
+        vm.host.scp(auth.auth, dest)
+
+        if use_xapi:
+            vm.host.ssh([
+                'varstore-set', vm.uuid, EFI_GUID_STRS[auth.name], auth.name,
+                str(EFI_AT_ATTRS), dest,
+            ])
+
+
+def generate_keys(self_signed=False):
+    PK = EFIAuth('PK')
+    KEK = EFIAuth('KEK')
+    db = EFIAuth('db')
+
+    if self_signed:
+        PK.sign_auth(PK)
+        KEK.sign_auth(KEK)
+        db.sign_auth(db)
+    else:
+        PK.sign_auth(PK)
+        PK.sign_auth(KEK)
+        KEK.sign_auth(db)
+
+    # For our tests the dbx blacklists anything signed by the db
+    dbx = EFIAuth.copy(db, name='dbx')
+
+    return PK, KEK, db, dbx
+
+
+class TestGuestLinuxUEFISecureBoot:
+    @pytest.fixture(autouse=True)
+    def setup_and_cleanup(self, linux_sb_vm):
+        vm = linux_sb_vm
+
+        self.PK, self.KEK, self.db, self.dbx = generate_keys()
+        vm.set_secureboot_enabled(False)
+        snapshot = vm.snapshot()
+        yield
+        snapshot.revert()
```

Comment:
In case of snapshot revert failure, one should still attempt to remove the messages, and only then fail. This comment applies generally to any teardown code that does more than one thing, by the way. I haven't checked every one of them.

Ideally, setup code that fails should also leave the system under test as clean as possible, but that's not something we've done systematically in other tests either.